clients/drcachesim/analyzer.cpp

```diff
@@ -431,50 +441,187 @@ analyzer_tmpl_t<RecordType, ReaderType>::process_tasks(analyzer_worker_data_t *w
+    assert(merged_intervals.empty());
+    // Used to recompute the interval_id for the result whole trace intervals, which are
+    // numbered by the earliest shard's timestamp.
+    uint64_t earliest_ever_interval_end_timestamp = std::numeric_limits<uint64_t>::max();
+    size_t shard_count = intervals.size();
+    bool any_shard_has_results_left = true;
+    while (any_shard_has_results_left) {
+        // Look for the next whole trace interval across all shards, which will be the
+        // one with the earliest interval-end timestamp.
+        uint64_t earliest_interval_end_timestamp = std::numeric_limits<uint64_t>::max();
+        for (size_t shard_idx = 0; shard_idx < shard_count; ++shard_idx) {
+            if (intervals[shard_idx].empty())
+                continue;
+            earliest_interval_end_timestamp =
+                std::min(earliest_interval_end_timestamp,
+                         intervals[shard_idx].front()->interval_end_timestamp_);
+        }
+        // We're done if no shard has any interval left unprocessed.
+        if (earliest_interval_end_timestamp == std::numeric_limits<uint64_t>::max()) {
+            any_shard_has_results_left = false;
+            continue;
+        }
+        assert(earliest_interval_end_timestamp % interval_microseconds_ == 0);
+        if (earliest_ever_interval_end_timestamp ==
+            std::numeric_limits<uint64_t>::max()) {
+            earliest_ever_interval_end_timestamp = earliest_interval_end_timestamp;
+        }
+
+        // Merge next intervals that have a timestamp == earliest_interval_end_timestamp.
+        // In other words: merge intervals from all shards that were active during this
+        // interval.
+        typename analysis_tool_tmpl_t<RecordType>::interval_state_snapshot_t
+            *cur_merged_interval = nullptr;
+        for (size_t shard_idx = 0; shard_idx < shard_count; ++shard_idx) {
+            if (intervals[shard_idx].empty())
+                continue;
+            uint64_t cur_interval_end_timestamp =
+                intervals[shard_idx].front()->interval_end_timestamp_;
+            assert(cur_interval_end_timestamp >= earliest_interval_end_timestamp);
+            if (cur_interval_end_timestamp > earliest_interval_end_timestamp)
+                continue;
+            // This shard was active during this interval. So, we merge the current
+            // shard's snapshot with the result.
+            if (cur_merged_interval == nullptr) {
+                cur_merged_interval = intervals[shard_idx].front();
+            } else {
+                typename analysis_tool_tmpl_t<RecordType>::interval_state_snapshot_t *res;
+                if (!merge_and_release_snapshots(tool_idx, res, cur_merged_interval,
+                                                 intervals[shard_idx].front()))
+                    return false;
+                cur_merged_interval = res;
+            }
+            intervals[shard_idx].pop();
```

Comment:
The merge logic needs FIFO, so a queue suits better imho.

>  it is all constructed up front with no insertions once created.

But are post-creation insertions really needed to justify use of a queue?

We need to process oldest interval first which is essentially FIFO. We never need to access some middle intervals arbitrarily; always the oldest.

> I would expect a vector with an index used here

Adding another array `next_interval_per_shard` to store next indices increases code size and complexity. Each access would become `A[i][B[i]]` instead of  a simple `A[i].front()`. We'd also need to add a separate loop at the end for releasing each snapshot.

> instead of using deletion to walk the container

The deletion is not to just walk the container. We really don't need the `pop()`d snapshot in the queue afterwards.

Reply:
Not really convinced; there's already a loop calling release.  At least add comments on why this is a queue.

Reply:
Will add the comment. I really believe a queue fits here, and additionally makes the code simpler.